File: packages/math/src/get_quat_directions.rs
use glam::{DQuat, DVec3};

#[derive(Debug, Clone)]
pub struct QuatDirections {
    pub up: DVec3,
    pub down: DVec3,
    pub left: DVec3,
    pub right: DVec3,
    pub forwards: DVec3,
    pub backwards: DVec3,
}

static FORWARDS: DVec3 = DVec3::new(0.0, 0.0, -1.0);
static BACKWARDS: DVec3 = DVec3::new(0.0, 0.0, 1.0);
static UP: DVec3 = DVec3::new(0.0, 1.0, 0.0);
static DOWN: DVec3 = DVec3::new(0.0, -1.0, 0.0);
static LEFT: DVec3 = DVec3::new(-1.0, 0.0, 0.0);
static RIGHT: DVec3 = DVec3::new(1.0, 0.0, 0.0);

pub fn get_quat_directions(quat: DQuat) -> QuatDirections {
    QuatDirections {
        forwards: quat * FORWARDS,
        backwards: quat * BACKWARDS,
        up: quat * UP,
        down: quat * DOWN,
        left: quat * LEFT,
        right: quat * RIGHT,
    }
}

File: game-script/framework/mathModule/getQuatDirections.ts
import { Quaternion, Vector3 } from "@aeroflightlabs/linear-math";

// TODO maybe move this to aflmath
export function getQuatDirections(q: Quaternion): {
  up: Vector3;
  down: Vector3;
  left: Vector3;
  right: Vector3;
  forwards: Vector3;
  backwards: Vector3;
} {
  const forwards = new Vector3(0, 0, -1).transformQuat(q);
  const backwards = new Vector3(0, 0, 1).transformQuat(q);
  const up = new Vector3(0, 1, 0).transformQuat(q);
  const down = new Vector3(0, -1, 0).transformQuat(q);
  const left = new Vector3(-1, 0, 0).transformQuat(q);
  const right = new Vector3(1, 0, 0).transformQuat(q);
  return {
    up,
    down,
    left,
    right,
    forwards,
    backwards,
  };
}
